trantor/net/inner/tlsprovider/BotanTLSProvider.cc

```diff
@@ -0,0 +1,390 @@
+            validationPolicy_.requireClientCert_ =
+                contextPtr_->requireClientCert;
+            channel_ = std::make_unique<Botan::TLS::Server>(
+                *this, sessionManager, *credsPtr_, validationPolicy_, rng);
+        }
+        else
+        {
+            validationPolicy_.requireClientCert_ =
+                contextPtr_->requireClientCert;
+            channel_ = std::make_unique<Botan::TLS::Client>(
+                *this,
+                sessionManager,
+                *credsPtr_,
+                validationPolicy_,
+                rng,
+                Botan::TLS::Server_Information(policyPtr_->getHostname(),
+                                               conn_->peerAddr().toPort()),
+                policyPtr_->getUseOldTLS()
+                    ? Botan::TLS::Protocol_Version::TLS_V10
+                    : Botan::TLS::Protocol_Version::TLS_V12,
+                policyPtr_->getAlpnProtocols());
+            setSniName(policyPtr_->getHostname());
+        }
+    }
+
+    void handleSSLError(SSLError err)
+    {
+        if (!errorCallback_)
+            return;
+        loop_->queueInLoop([this, err]() { errorCallback_(conn_, err); });
+    }
+
+    virtual ~BotanTLSProvider() override = default;
+
+    void tls_emit_data(const uint8_t data[], size_t size) override
+    {
+        auto n = writeCallback_(conn_, data, size);
+        lastWriteSize_ = n;
+
+        // store the unsent data and send it later
+        if (n == ssize_t(size))
+            return;
+        if (n == -1)
+            n = 0;
+        appendToWriteBuffer((const char *)data + n, size - n);
+    }
+
+    void tls_record_received(uint64_t seq_no,
+                             const uint8_t data[],
+                             size_t size) override
+    {
+        (void)seq_no;
+        recvBuffer_.append((const char *)data, size);
+        if (messageCallback_)
+            messageCallback_(conn_, &recvBuffer_);
+    }
+
+    void tls_alert(Botan::TLS::Alert alert) override
+    {
+        if (alert.type() == Botan::TLS::Alert::CLOSE_NOTIFY)
+        {
+            LOG_WARN << "TLS close notify received";
+            if (closeCallback_)
+                closeCallback_(conn_);
+        }
+        else
+        {
+            if (errorCallback_)
+                errorCallback_(conn_, SSLError::kSSLProtocolError);
+        }
+    }
+
+    bool tls_session_established(const Botan::TLS::Session &session)
+    {
+        LOG_TRACE << "tls_session_established";
+        bool needCerts =
+            policyPtr_->getValidateDate() || policyPtr_->getValidateDomain();
+        const auto &certs = session.peer_certs();
+        if (needCerts && session.peer_certs().empty())
+        {
+            handleSSLError(SSLError::kSSLInvalidCertificate);
+            throw Botan::TLS::TLS_Exception(Botan::TLS::Alert::NO_CERTIFICATE,
+                                            "No certificate received");
+        }
```

Comment:
I don't think you need to check that here, as the default implementation of `tls_verify_cert_chain` will check that already and throw a `TLS_Exception` if no certificates were sent by the peer. See the other, similar, comments about how to handle this case.